File: day2/python/part2.py
with open("../input.txt") as file:
    instructions = file.read().splitlines();
    
    
pos = [0,0]
aim = 0
for ins in instructions:
    ins = ins.split(" ")
    direction = ins[0]
    amount = int(ins[1])
    if direction == "forward":
        pos[0] += amount
        pos[1] += aim * amount
    if direction == "down":
        aim += amount
    if direction == "up":
        aim -= amount
        
print(pos[0]*pos[1])

File: day2/haskell/part2.hs
module Main where
import Data.Char (isAlpha)

main :: IO ()
main = do
    entries <- map (span isAlpha) <$> (lines <$> readFile "../input.txt")
    let pos = ((0,0),0)
    print $ uncurry (*) $ fst $ foldl (flip increase) pos entries

increase :: (String, String) -> ((Integer, Integer), Integer) -> ((Integer, Integer), Integer)
increase (direction, n) ((x,y), aim) =
    let amount = read n :: Integer in
    case direction of
        "forward" -> ((x+amount,y+(amount*aim)),aim)
        "down"    -> ((x,y),aim+amount)
        "up"      -> ((x,y),aim-amount)
        _         -> error "invalid insturction"